cpp/src/arrow/extension_type-test.cc

```diff
@@ -0,0 +1,328 @@
+    if (serialized != "uuid-type-unique-code") {
+      return Status::Invalid("Type identifier did not match");
+    }
+    DCHECK(storage_type->Equals(*fixed_size_binary(16)));
+    *out = std::make_shared<UUIDType>();
+    return Status::OK();
+  }
+
+  std::string Serialize() const override { return "uuid-type-unique-code"; }
+};
+
+std::shared_ptr<DataType> uuid() { return std::make_shared<UUIDType>(); }
+
+class Parametric1Array : public ExtensionArray {
+ public:
+  using ExtensionArray::ExtensionArray;
+};
+
+class Parametric2Array : public ExtensionArray {
+ public:
+  using ExtensionArray::ExtensionArray;
+};
+
+// A parametric type where the extension_name() is always the same
+class Parametric1Type : public ExtensionType {
+ public:
+  explicit Parametric1Type(int32_t parameter)
+      : ExtensionType(int32()), parameter_(parameter) {}
+
+  int32_t parameter() const { return parameter_; }
+
+  std::string extension_name() const override { return "parametric-type-1"; }
+
+  bool ExtensionEquals(const ExtensionType& other) const override {
+    const auto& other_ext = static_cast<const ExtensionType&>(other);
+    if (other_ext.extension_name() != this->extension_name()) {
+      return false;
+    }
+    return this->parameter() == static_cast<const Parametric1Type&>(other).parameter();
+  }
+
+  std::shared_ptr<Array> MakeArray(std::shared_ptr<ArrayData> data) const override {
+    return std::make_shared<Parametric1Array>(data);
+  }
+
+  Status Deserialize(std::shared_ptr<DataType> storage_type,
+                     const std::string& serialized,
+                     std::shared_ptr<DataType>* out) const override {
+    DCHECK_EQ(4, serialized.size());
+    const int32_t parameter = *reinterpret_cast<const int32_t*>(serialized.data());
+    DCHECK(storage_type->Equals(int32()));
+    *out = std::make_shared<Parametric1Type>(parameter);
+    return Status::OK();
+  }
+
+  std::string Serialize() const override {
+    std::string result("    ");
+    memcpy(&result[0], &parameter_, sizeof(int32_t));
+    return result;
+  }
+
+ private:
+  int32_t parameter_;
+};
+
+// A parametric type where the extension_name() is different for each
+// parameter, and must be separately registered
+class Parametric2Type : public ExtensionType {
+ public:
+  explicit Parametric2Type(int32_t parameter)
+      : ExtensionType(int32()), parameter_(parameter) {}
+
+  int32_t parameter() const { return parameter_; }
+
+  std::string extension_name() const override {
+    std::stringstream ss;
+    ss << "parametric-type-2<param=" << parameter_ << ">";
+    return ss.str();
+  }
+
+  bool ExtensionEquals(const ExtensionType& other) const override {
+    const auto& other_ext = static_cast<const ExtensionType&>(other);
+    if (other_ext.extension_name() != this->extension_name()) {
+      return false;
+    }
+    return this->parameter() == static_cast<const Parametric2Type&>(other).parameter();
```

Comment:
This is redundant with checking the extension name, right?

Reply:
in this case yes